File: src/managers/objectmanager.ts
// Port of DAL/objectmanager.{h,cpp}.

import { Coordinate } from '../core/coordinate';
import { ImageVectors } from '../core/images';
import { IObjectManager, IGameScene, IGameSettingsManager } from '../model/base';
import { TileBase } from '../model/tile';
import { HeadQuarters } from '../model/building';
import { PlayerBase } from '../model/player';
import { ClickedTileBorder, BlockedTile, MouseHoverBorder } from '../model/overlays';
import type { IGameEventHandler } from '../model/base';

interface ICurrentPlayerSource {
  getCurrentPlayer(): PlayerBase;
}

export class ObjectManager implements IObjectManager {
  private tiles_: TileBase[] = [];
  private hoverBorder_: MouseHoverBorder | null = null;
  private clickedTileBorder_: ClickedTileBorder | null = null;
  private blockedTileOverlays_: BlockedTile[] = [];
  private gameScene_!: IGameScene;
  private gameEventHandler_!: IGameEventHandler & ICurrentPlayerSource;
  private gameSettingsManager_!: IGameSettingsManager;

  setGameScene(gs: IGameScene): void {
    this.gameScene_ = gs;
  }
  getGameScene(): IGameScene {
    return this.gameScene_;
  }

  addDALS(
    gameeventhandler: IGameEventHandler & ICurrentPlayerSource,
    _menuobjectmanager: unknown,
    gamesettingsmanager: IGameSettingsManager,
  ): void {
    this.gameEventHandler_ = gameeventhandler;
    this.gameSettingsManager_ = gamesettingsmanager;
  }

  addTiles(tiles: TileBase[]): void {
    for (const tile of tiles) this.tiles_.push(tile);
  }

  replaceTile(oldTile: TileBase, newTile: TileBase): void {
    const idx = this.tiles_.indexOf(oldTile);
    if (idx !== -1) {
      this.tiles_.splice(idx, 1);
      this.tiles_.push(newTile);
      newTile.setOwner(oldTile.getOwner());
      for (const unit of oldTile.getUnits()) {
        unit.addParentTile(newTile);
        newTile.addUnit(unit);
      }
    } else {
      console.warn('Error, tile to be replaced was not found.');
    }
  }

  getTile(coordinate: Coordinate): TileBase | null {
    for (const tile of this.tiles_) {
      if (tile.getCoordinate().equals(coordinate)) return tile;
    }
    return null;
  }

  getTiles(): TileBase[] {
    return this.tiles_;
  }

  setHoverBorder(border: MouseHoverBorder): void {
    this.hoverBorder_ = border;
  }

  setClickedTileBorder(tile: TileBase): void {
    this.clickedTileBorder_ = new ClickedTileBorder(
      tile.getCoordinate(),
      1,
      1,
      this.gameEventHandler_,
      this,
    );
    this.clickedTileBorder_.setImageFiles(ImageVectors.CLICKEDTILEBORDER);
    this.gameScene_.drawItem(this.clickedTileBorder_);
  }

  getClickedTileBorder(): ClickedTileBorder | null {
    return this.clickedTileBorder_;
  }

  removeClickedTileBorder(): void {
    if (this.clickedTileBorder_ !== null) {
      this.gameScene_.removeItem(this.clickedTileBorder_);
    }
    this.clickedTileBorder_ = null;
  }

  getBorderTile(): MouseHoverBorder | null {
    return this.hoverBorder_;
  }

  getHqConnectedTiles(player: PlayerBase): TileBase[] {
    const tiles: TileBase[] = [];
    const hq = this.getHqTile(player);
    if (hq !== null) {
      tiles.push(hq);
      for (let i = 0; i < tiles.length; i++) {
        for (const neighbour of tiles[i].getNeighbourFourTiles()) {
          if (tiles.includes(neighbour)) continue;
          if (player === neighbour.getOwner()) tiles.push(neighbour);
        }
      }
    }
    return tiles;
  }

  getHqTile(player: PlayerBase): TileBase | null {
    for (const object of player.getObjects()) {
      if (object instanceof TileBase) {
        const building = object.getBuilding();
        if (building !== null && building.getType() === 'Headquarters') {
          if (!(building as HeadQuarters).isConquered()) {
            return object;
          }
        }
      }
    }
    return null;
  }

  getAvailableTiles(): TileBase[] {
    const player = this.gameEventHandler_.getCurrentPlayer();
    const availableTiles: TileBase[] = [];

    for (const obj of player.getObjects()) {
      if (obj instanceof TileBase) {
        const tile = obj;
        if (tile.getOwner() === player && tile.hasOpponentHeadquarters(player)) {
          if (!availableTiles.includes(tile)) availableTiles.push(tile);
        }

        if (tile.getType() === 'River' && tile.getBuilding() === null) {
          continue;
        } else {
          for (const nTile of tile.getNeighbourFourTiles()) {
            if (availableTiles.includes(nTile)) continue;
            if (nTile.hasOpponentHeadquarters(player)) availableTiles.push(nTile);
          }
        }
      }
    }
    return availableTiles;
  }

  addBlockTileOverlays(): void {
    const availableTiles = this.getAvailableTiles();
    const blockedTiles: TileBase[] = [];
    for (const tile of this.getTiles()) {
      if (!availableTiles.includes(tile)) blockedTiles.push(tile);
    }
    for (const blockedTile of blockedTiles) {
      const overlay = new BlockedTile(blockedTile.getCoordinate(), 1, 1, this.gameEventHandler_, this);
      overlay.setImageFiles(ImageVectors.BLOCKED_TILE);
      this.blockedTileOverlays_.push(overlay);
      this.gameScene_.drawItem(overlay);
    }
  }

  removeBlockTileOverlays(): void {
    for (const overlay of this.blockedTileOverlays_) {
      this.gameScene_.removeItem(overlay);
    }
    this.blockedTileOverlays_ = [];
  }

  /** The tile carrying the single Strange Device, if one exists (the one-per-game rule). */
  findStrangeDeviceTile(): TileBase | null {
    for (const tile of this.tiles_) {
      const b = tile.getBuilding();
      if (b !== null && b.getType() === 'Strange Device') return tile;
    }
    return null;
  }

  hasStrangeDevice(): boolean {
    return this.findStrangeDeviceTile() !== null;
  }

  getTileCount(): number {
    return this.tiles_.length;
  }

  getTileCountForPlayer(player: PlayerBase): number {
    let n = 0;
    for (const tile of this.tiles_) if (tile.getOwner() === player) n++;
    return n;
  }

  getNeutralTiles(): number {
    let n = 0;
    for (const tile of this.tiles_) if (tile.getOwner() === null) n++;
    return n;
  }
}

File: reference/Game/DAL/objectmanager.cpp
/*
######################################################################
# TIE-02402 Ohjelmointi 3: Perusteet, S2019                          #
#                                                                    #
# Project: Colonizing Pirkanmaa                                      #
# Program description: Program instructions are located in           #
#                      Documentation/documentation.pdf               #
#                                                                    #
# File: objectmanager.cpp, see objectmanager.h for more info         #
#                                                                    #
# Authors: Otto Ranta-Ojala, 253561, otto.ranta-ojala@tuni.fi        #
#          Tino Kaartovuori, 254987, tino.kaartovuori@tuni.fi        #
######################################################################
*/

#include "objectmanager.h"
#include "iostream"
#include <QDebug>

namespace Student {

ObjectManager::ObjectManager()
{
}


void ObjectManager::setGameScene(std::shared_ptr<GameScene> gs)
{
    gameScene_ = gs;
}


std::shared_ptr<GameScene> ObjectManager::getGameScene()
{
    return gameScene_.lock();
}


void ObjectManager::addTiles(
        const std::vector<std::shared_ptr<Course::TileBase>> &tiles)
{
    for (const auto& tile : tiles) {
        tiles_.push_back(tile);
    }
}

void ObjectManager::replaceTile(std::shared_ptr<Course::TileBase> oldTile,
                               std::shared_ptr<Course::TileBase> newTile)
{
    bool removed = false;
    for(std::vector<std::shared_ptr<Course::TileBase>>::iterator
                              it = tiles_.begin(); it != tiles_.end();)
    {
        if (*it == oldTile){
            it = tiles_.erase(it);
            removed = true;
            break;
        } else {
            ++it;
        }
    }

    if (removed) {
        tiles_.push_back(newTile);
        newTile->setOwner(oldTile->getOwner());
        for (auto unit : oldTile->getUnits()) {
            unit->addParentTile(newTile);
            newTile->addUnit(unit);
        }
    }
    else {
        qDebug()<<"Error, tile to be replaced was not found.";
    }
}


std::shared_ptr<Course::TileBase> ObjectManager::getTile
                                (const Course::Coordinate &coordinate)
{
    for (const auto& tile : tiles_) {
        if (tile->getCoordinate() == coordinate) {
            return tile;
        }
    }

    return nullptr; //If the tile wasn't found a nullptr will be returned
}


std::vector<std::shared_ptr<Course::TileBase> > ObjectManager::getTiles()
{
    return tiles_;
}


void ObjectManager::setHoverBorder(
        const std::shared_ptr<Student::MouseHoverBorder> border)
{
    hoverBorder_ = border;
}


void ObjectManager::setClickedTileBorder(std::shared_ptr<Course::TileBase> tile)
{
    clickedTileBorder_ = std::make_shared<Student::ClickedTileBorder>
            (tile->getCoordinate(), 1, 1,
             gameEventHandler_.lock(), shared_from_this());

   clickedTileBorder_->setImageFiles(ImageVectors::CLICKEDTILEBORDER);
   gameScene_.lock()->drawItem(clickedTileBorder_);

}


std::shared_ptr<Student::ClickedTileBorder> ObjectManager::getClickedTileBorder()
{
    return clickedTileBorder_;
}


void ObjectManager::removeClickedTileBorder()
{
    if (clickedTileBorder_ != nullptr) {
        gameScene_.lock()->removeItem(clickedTileBorder_);
    }

    clickedTileBorder_ = nullptr;
}


std::shared_ptr<Student::MouseHoverBorder> ObjectManager::getBorderTile()
{
    return hoverBorder_;
}


void ObjectManager::addDALS(
        const std::shared_ptr<Course::iGameEventHandler> gameeventhandler,
        const std::shared_ptr<Student::iMenuObjectManager> menuobjectmanager,
        const std::shared_ptr<Student::GameSettingsManager> gamesettingsmanager)
{
    gameEventHandler_ = gameeventhandler;
    menuObjectManager_ = menuobjectmanager;
    gameSettingsManager_ = gamesettingsmanager;
}


std::vector<std::shared_ptr<Course::TileBase>> ObjectManager::getHqConnectedTiles
                                     (std::shared_ptr<Course::PlayerBase> player)
{
    std::vector<std::shared_ptr<Course::TileBase>> tiles;

    //Headquarters exist
    if (getHqTile(player) != nullptr) {
        tiles.push_back(getHqTile(player));

        //When all tiles have been found the for-loop ends
        for (int i = 0; i < (int)tiles.size(); ++i) {
            /*Loops all neighbouring tiles and adds them into tiles-vector
             *if the player owns the tile*/
            for (auto neighbour: tiles.at(i)->getNeighbourFourTiles())
            {
                /*If the neighbour has already been added it will be skipped
                 *and won't be added*/
                if (std::find(tiles.begin(), tiles.end(),
                              neighbour) != tiles.end()) {
                    continue;
                }
                if (player == neighbour->getOwner()) {
                    tiles.push_back(neighbour);
                }
            }
        }
    }

    return tiles;
}


std::shared_ptr<Course::TileBase> ObjectManager::getHqTile
                            (std::shared_ptr<Course::PlayerBase> player)
{
    std::shared_ptr<Course::TileBase> HqTile;

    //Loops all player's objects
    for (auto object : player->getObjects()) {
        if (std::dynamic_pointer_cast<Course::TileBase>(object) != nullptr)
        {
            std::shared_ptr<Course::TileBase> tile =
                    std::dynamic_pointer_cast<Course::TileBase>(object);

            if (tile->getBuilding() != nullptr &&
                    tile->getBuilding()->getType() == "Headquarters")
            {
                //Tile with headquarters is found
                std::shared_ptr<Course::HeadQuarters> HQ =
                                std::dynamic_pointer_cast
                                <Course::HeadQuarters>(tile->getBuilding());
                //Only non-conquered headquarters tile is returned.
                //A conquered one doesn't count
                if (!HQ->isConquered()) {
                    HqTile = tile;
                    break;
                }
            }
        }
    }

    return HqTile;
}


std::vector<std::shared_ptr<Course::TileBase>> ObjectManager::getAvailableTiles()
{
    std::shared_ptr<Course::PlayerBase> player = gameEventHandler_.lock()
                                                    ->getCurrentPlayer();
    std::vector<std::shared_ptr<Course::TileBase>> availableTiles = {};

    for (auto obj : player->getObjects()) {
        if (std::dynamic_pointer_cast<Course::TileBase>(obj) != nullptr) {
            std::shared_ptr<Course::TileBase> tile =
                                   std::dynamic_pointer_cast<Course::TileBase>(obj);

            // Check if tile is already in the available tiles

            // Tile is owned by player and units can be placed
            if (tile->getOwner() == player and tile
                    ->hasOpponentHeadquarters(player))
            {
                if (std::find(availableTiles.begin(), availableTiles.end(), tile)
                                                        != availableTiles.end()) {
                    // tile is already in available tiles
                } else {
                    availableTiles.push_back(tile);
                }
            }

            if (tile->getType() == "River" and tile->getBuilding() == nullptr) {
                continue;
            } else {
                std::vector<std::shared_ptr<Course::TileBase>> neighbourTiles =
                        tile->getNeighbourFourTiles();
                for (std::shared_ptr<Course::TileBase> n_tile : neighbourTiles) {
                    if (std::find(availableTiles.begin(), availableTiles.end(),
                                  n_tile) != availableTiles.end()) {
                        // tile is already in available tiles
                        continue;
                    } else {
                        if (n_tile->hasOpponentHeadquarters(player)) {
                            availableTiles.push_back(n_tile);
                        }
                    }
                }
            }
        }
    }

    return availableTiles;
}


void ObjectManager::addBlockTileOverlays() {
    std::vector<std::shared_ptr<Course::TileBase>> availableTiles =
                                    getAvailableTiles();
    std::vector<std::shared_ptr<Course::TileBase>> blockedTiles = {};

    for (auto tile : getTiles()) {
        bool tileFound = false;
        for (auto a_tile : availableTiles) {
            if (tile == a_tile) {
                tileFound = true;
                break;
            }
        }
        if (!tileFound) {
            blockedTiles.push_back(tile);
        }
    }

    for (auto blockedTile : blockedTiles) {
        std::shared_ptr<Student::BlockedTile> overlay =
                std::make_shared<Student::BlockedTile>(
                                            blockedTile->getCoordinate(),
                                            1,
                                            1,
                                            gameEventHandler_,
                                            shared_from_this());
        overlay->setImageFiles(ImageVectors::BLOCKED_TILE);
        blockedTileOverlays_.push_back(overlay);
        gameScene_.lock()->drawItem(overlay);
    }
}


void ObjectManager::removeBlockTileOverlays() {

    for (auto overlay : blockedTileOverlays_) {
        gameScene_.lock()->removeItem(overlay);
    }

    blockedTileOverlays_.clear();
}


int ObjectManager::getTileCount()
{
    return tiles_.size();
}


int ObjectManager::getTileCountForPlayer
                        (std::shared_ptr<Course::PlayerBase> player)
{
    int tileCount = 0;
    for (auto tile : tiles_) {
        if (tile->getOwner() == player) {
            ++tileCount;
        }
    }
    return tileCount;
}


int ObjectManager::getNeutralTiles()
{
    int tileCount = 0;
    for (auto tile : tiles_) {
        if (tile->getOwner() == nullptr) {
            ++tileCount;
        }
    }
    return tileCount;
}


} //Namespace Student


